plutus-pab/src/Plutus/PAB/Effects/Contract/ContractExe.hs

```diff
@@ -0,0 +1,173 @@
+    type State ContractExe = PartiallyDecodedResponse ContractPABRequest
+
+    serialisableState _ = id
+
+newtype ContractExe =
+    ContractExe
+        { contractPath :: FilePath
+        }
+    deriving (Show, Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+instance StructuredLog ContractExe where
+    toStructuredLog e = HM.singleton "contract" (toJSON e)
+
+instance Pretty ContractExe where
+    pretty ContractExe {contractPath} = "Path:" <+> pretty contractPath
+
+-- | Handle 'ContractEffect ContractExe' by making calls to compiled contract
+--   executables
+handleContractEffectContractExe ::
+       ( Member (LogMsg ContractExeLogMsg) effs
+       , Member (Error PABError) effs
+       , LastMember m effs
+       , MonadIO m)
+    => ContractEffect ContractExe
+    ~> Eff effs
+handleContractEffectContractExe =
+    \case
+        InitialState (ContractExe contractPath) -> do
+            logDebug $ InitContractMsg contractPath
+            liftProcess $ readProcessWithExitCode contractPath ["init"] ""
+        UpdateContract (ContractExe contractPath) (oldState :: PartiallyDecodedResponse ContractPABRequest) (input :: Response Events.Contract.ContractResponse) -> do
+            let req :: ContractRequest Value
+                req = ContractRequest{oldState = ContractInstanceState.newState oldState, event = toJSON <$> input}
+                pl = BSL8.unpack (JSON.encodePretty req)
+            logDebug $ UpdateContractMsg contractPath req
+            liftProcess $ readProcessWithExitCode contractPath ["update"] pl
+        ExportSchema (ContractExe contractPath) -> do
+            logDebug $ ExportSignatureMsg contractPath
+            liftProcess $
+                readProcessWithExitCode contractPath ["export-signature"] ""
+
+liftProcess ::
+       (LastMember m effs, MonadIO m, FromJSON b, Member (LogMsg ContractExeLogMsg) effs, Member (Error PABError) effs)
+    => IO (ExitCode, String, String)
+    -> Eff effs b
+liftProcess process = do
+    (exitCode, stdout, stderr) <- sendM $ liftIO process
+    case exitCode of
+        ExitFailure code -> do
+            logDebug $ ProcessExitFailure stderr
+            throwError $ ContractCommandError code (Text.pack stderr)
+        ExitSuccess -> do
+            logDebug $ ContractResponse stdout
+            case JSON.eitherDecode (BSL8.pack stdout) of
+                Right value -> pure value
+                Left err    -> throwError $ ContractCommandError 0 (Text.pack err)
+
+data ContractExeLogMsg =
+    InvokeContractMsg
+    | InitContractMsg FilePath
+    | UpdateContractMsg FilePath (ContractRequest Value)
+    | ExportSignatureMsg FilePath
+    | ProcessExitFailure String
+    | ContractResponse String
+    | Migrating
+    | InvokingEndpoint String Value
+    | EndpointInvocationResponse [Text]
+    | ContractExePABError PABError
+    deriving stock (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty ContractExeLogMsg where
```

Comment:
Seeing all the new/changed msg data types and respective instances for logging added i'd like to mention something that I pointed out in some chat before: I don't think we'll ever get consistent logging with all of those. Too many different instances in too many different places for that.

I'm also not convinced it is useful or helpful to be able to track what is being logged where in the effect system (while this on the other hand leads to rather tight coupling of types).

This PR is far too huge already but I think there should be _one_  logging data type with some mandatory fields (like say `component` and `isMock` and maybe some others i can't think of now) and some free-text field.

Reply:
Yeah I agree with you. We should keep the structured logging in the emulator (`EmulatorLogMsg`) but we don't really need it here in the PAB.